code/modules/projectiles/gun.dm

```diff
@@ -527,34 +527,37 @@
-			return
-		return remove_gun_attachment(user, I, item_to_remove)
-
-	else if(gun_light && can_flashlight) //if it has a gun_light and can_flashlight is false, the flashlight is permanently attached.
-		return remove_gun_attachment(user, I, gun_light, "unscrewed")
-
-	else if(bayonet && can_bayonet) //if it has a bayonet, and the bayonet can be removed
-		return remove_gun_attachment(user, I, bayonet, "unfix")
-
-	else if(pin && user.is_holding(src))
-		user.visible_message(span_warning("[user] attempts to remove [pin] from [src] with [I]."),
-		span_notice("You attempt to remove [pin] from [src]. (It will take [DisplayTimeText(FIRING_PIN_REMOVAL_DELAY)].)"), null, 3)
-		if(I.use_tool(src, user, FIRING_PIN_REMOVAL_DELAY, volume = 50))
-			if(!pin) //check to see if the pin is still there, or we can spam messages by clicking multiple times during the tool delay
-				return
-			user.visible_message(span_notice("[pin] is pried out of [src] by [user], destroying the pin in the process."),
-								span_warning("You pry [pin] out with [I], destroying the pin in the process."), null, 3)
-			QDEL_NULL(pin)
-			return TRUE
+    . = ..()
+    if(.)
+        return
+    if(!user.canUseTopic(src, BE_CLOSE, FALSE, NO_TK))
+        return
+    if((can_flashlight && gun_light) && (can_bayonet && bayonet)) //give them a choice instead of removing both
+        var/list/possible_items = list(gun_light, bayonet)
+        var/obj/item/item_to_remove = input(user, "Select an attachment to remove", "Attachment Removal") as null|obj in possible_items
+        if(!item_to_remove || !user.canUseTopic(src, BE_CLOSE, FALSE, NO_TK))
+            return
+        return remove_gun_attachment(user, I, item_to_remove)
+
+    else if(gun_light && can_flashlight) //if it has a gun_light and can_flashlight is false, the flashlight is permanently attached.
+        return remove_gun_attachment(user, I, gun_light, "unscrewed")
+
+    else if(bayonet && can_bayonet) //if it has a bayonet, and the bayonet can be removed
+        return remove_gun_attachment(user, I, bayonet, "unfix")
+
+    else if(pin && user.is_holding(src))
+
+        if (!pin.pin_removeable)
+            return
+        user.visible_message(span_warning("[user] attempts to remove [pin] from [src] with [I]."),
+        span_notice("You attempt to remove [pin] from [src]. (It will take [DisplayTimeText(FIRING_PIN_REMOVAL_DELAY)].)"), null, 3)
+        if(I.use_tool(src, user, FIRING_PIN_REMOVAL_DELAY, volume = 50))
+            if(!pin) //check to see if the pin is still there, or we can spam messages by clicking multiple times during the tool delay
+                return
+            user.visible_message(span_notice("[pin] is pried out of [src] by [user], destroying the pin in the process."),
+                                span_warning("You pry [pin] out with [I], destroying the pin in the process."), null, 3)
+            QDEL_NULL(pin)
+            return TRUE
```

Comment:
Please use tabs instead of spaces to indent code, check the rest of the pr in case this is also the case for any other indenting

Reply:
Fixed